File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/NeighborList.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.net.Inet4Address;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * SpaceServer to Simulator Messages
 * Neighbor List - Passed anytime neighbors change
 *
 * <p>Template: {@code NeighborList High 3 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class NeighborList extends SLMessage {
    public NeighborBlock[] NeighborBlock_Fields = new NeighborBlock[4];

    /** Block NeighborBlock, Multiple 4. */
    public static class NeighborBlock {
        public Inet4Address IP; // IPADDR
        public byte[] Name; // Variable 1 - string
        public int Port; // IPPORT
        public Inet4Address PublicIP; // IPADDR
        public int PublicPort; // IPPORT
        public UUID RegionID; // LLUUID
        public int SimAccess; // U8
    }

    public NeighborList() {
        this.zeroCoded = false;
        for (int i = 0; i < 4; i++) {
            this.NeighborBlock_Fields[i] = new NeighborBlock();
        }
    }

    @Override
    public int CalcPayloadSize() {
        int i = 1;
        for (int i2 = 0; i2 < 4; i2++) {
            i += this.NeighborBlock_Fields[i2].Name.length + 29 + 1;
        }
        return i;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleNeighborList(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: High 3 (NeighborList).
        byteBuffer.put((byte) 0x03);
        for (int i = 0; i < 4; i++) {
            packIPAddress(byteBuffer, this.NeighborBlock_Fields[i].IP);
            packShort(byteBuffer, (short) this.NeighborBlock_Fields[i].Port);
            packIPAddress(byteBuffer, this.NeighborBlock_Fields[i].PublicIP);
            packShort(byteBuffer, (short) this.NeighborBlock_Fields[i].PublicPort);
            packUUID(byteBuffer, this.NeighborBlock_Fields[i].RegionID);
            packVariable(byteBuffer, this.NeighborBlock_Fields[i].Name, 1);
            packByte(byteBuffer, (byte) this.NeighborBlock_Fields[i].SimAccess);
        }
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        for (int i = 0; i < 4; i++) {
            this.NeighborBlock_Fields[i].IP = unpackIPAddress(byteBuffer);
            this.NeighborBlock_Fields[i].Port = unpackShort(byteBuffer) & 65535;
            this.NeighborBlock_Fields[i].PublicIP = unpackIPAddress(byteBuffer);
            this.NeighborBlock_Fields[i].PublicPort = unpackShort(byteBuffer) & 65535;
            this.NeighborBlock_Fields[i].RegionID = unpackUUID(byteBuffer);
            this.NeighborBlock_Fields[i].Name = unpackVariable(byteBuffer, 1);
            this.NeighborBlock_Fields[i].SimAccess = unpackByte(byteBuffer) & 0xFF;
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/NeighborList.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.net.Inet4Address
import java.nio.ByteBuffer
import java.util.UUID

/**
 * SpaceServer to Simulator Messages
 * Neighbor List - Passed anytime neighbors change
 *
 * <p>Template: {@code NeighborList High 3 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class NeighborList : SLMessage() {
    @JvmField val NeighborBlock_Fields = Array(4) { NeighborBlock() }

    /** Block NeighborBlock, Multiple 4. */
    open class NeighborBlock {
        @JvmField var IP: Inet4Address? = null
        @JvmField var Name: ByteArray? = null
        @JvmField var Port: Int = 0
        @JvmField var PublicIP: Inet4Address? = null
        @JvmField var PublicPort: Int = 0
        @JvmField var RegionID: UUID? = null
        @JvmField var SimAccess: Int = 0
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        var i = 1
        (0 until 4).forEach { index ->
            i += NeighborBlock_Fields[index].Name!!.size + 29 + 1
        }
        return i
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleNeighborList(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: High 3 (NeighborList).
        byteBuffer.put((0x03).toByte())
        (0 until 4).forEach { index ->
            packIPAddress(byteBuffer, NeighborBlock_Fields[index].IP)
            packShort(byteBuffer, NeighborBlock_Fields[index].Port.toShort())
            packIPAddress(byteBuffer, NeighborBlock_Fields[index].PublicIP)
            packShort(byteBuffer, NeighborBlock_Fields[index].PublicPort.toShort())
            packUUID(byteBuffer, NeighborBlock_Fields[index].RegionID)
            packVariable(byteBuffer, NeighborBlock_Fields[index].Name, 1)
            packByte(byteBuffer, (NeighborBlock_Fields[index].SimAccess).toByte())
        }
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        (0 until 4).forEach { index ->
            NeighborBlock_Fields[index].IP = unpackIPAddress(byteBuffer)
            NeighborBlock_Fields[index].Port = unpackShort(byteBuffer).toInt() and 65535
            NeighborBlock_Fields[index].PublicIP = unpackIPAddress(byteBuffer)
            NeighborBlock_Fields[index].PublicPort = unpackShort(byteBuffer).toInt() and 65535
            NeighborBlock_Fields[index].RegionID = unpackUUID(byteBuffer)
            NeighborBlock_Fields[index].Name = unpackVariable(byteBuffer, 1)
            NeighborBlock_Fields[index].SimAccess = unpackByte(byteBuffer).toInt() and 0xFF
        }
    }
}
